src/circuit_builder.rs

```diff
@@ -26,21 +26,71 @@ pub enum PublicInput {
 }
 
 impl PublicInput {
-    /// Returns the value of a PublicInput struct
-    pub fn value(&self) -> Vec<BlsScalar> {
+    /// Returns the serialized-size of the `PublicInput` structure.
+    pub const fn serialized_size() -> usize {
+        33usize
+    }
+
+    /// Returns the byte-representation of a [`PublicInput`].
+    /// Note that the underlying variants of this enum have different
+    /// sizes on it's byte-representation. Therefore, we need to return
+    /// the biggest one to set it as the default one.
+    pub fn to_bytes(&self) -> [u8; Self::serialized_size()] {
+        let mut bytes = [0u8; Self::serialized_size()];
         match self {
-            PublicInput::BlsScalar(scalar, _) => vec![*scalar],
-            PublicInput::JubJubScalar(scalar, _) => vec![BlsScalar::from(*scalar)],
-            PublicInput::AffinePoint(point, _, _) => vec![point.get_x(), point.get_y()],
+            Self::BlsScalar(scalar, _) => {
+                bytes[0] = 0;
+                bytes[1..33].copy_from_slice(&scalar.to_bytes());
+                bytes
+            }
+            Self::JubJubScalar(scalar, _) => {
+                bytes[0] = 1;
+                bytes[1..33].copy_from_slice(&scalar.to_bytes());
+                bytes
+            }
+            Self::AffinePoint(point, _, _) => {
+                bytes[0] = 2;
+                bytes[1..Self::serialized_size()].copy_from_slice(&point.to_bytes());
+                bytes
+            }
+        }
+    }
+
+    /// Generate a [`PublicInput`] structure from it's byte representation.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, CircuitErrors> {
+        if bytes.len() != Self::serialized_size() {
```

Comment:
It should be `<`: you can always pass a bigger slice if you want to, maybe because it's part of a bigger buffer; the important thing is that you have enough bytes to parse and they're properly encoded.